js/private/js_library.bzl

```diff
@@ -68,48 +83,84 @@ target in {file_basename}'s package and add that target to the deps of {this_tar
 
+    sources = []
+    declarations = []
+
     for src in output_srcs:
         if src.is_directory:
             # assume a directory contains typings since we can't know that it doesn't
-            typings.append(src)
+            declarations.append(src)
+            sources.append(src)
         elif (
             src.path.endswith(".d.ts") or
-            src.path.endswith(".d.ts.map") or
-            # package.json may be required to resolve "typings" key
-            src.path.endswith("/package.json")
+            src.path.endswith(".d.ts.map")
         ):
-            typings.append(src)
-
-    files_depsets = [depset(output_srcs)]
-
-    for dep in ctx.attr.deps:
-        if DefaultInfo in dep:
-            files_depsets.append(dep[DefaultInfo].files)
+            declarations.append(src)
+        elif src.path.endswith("/package.json"):
+            # package.json may be required to resolve "typings" key
+            declarations.append(src)
+            sources.append(src)
+        else:
+            sources.append(src)
+
+    sources_depset = depset(sources)
+    transitive_sources_depset = depset(
+        direct = sources,
+        transitive = [
+            target[JsInfo].transitive_sources
+            for target in ctx.attr.deps
+            if JsInfo in target
+        ],
+    )
 
-    runfiles = ctx.runfiles(
-        files = output_srcs,
-        # We do not include typings_depsets in the runfiles because that would cause type-check actions to occur
-        # in every development workflow.
-        transitive_files = depset(transitive = files_depsets),
+    declarations_depset = depset(declarations)
+    transitive_declarations_depset = depset(
+        direct = declarations,
+        transitive = [
+            target[DeclarationInfo].transitive_declarations
+            for target in ctx.attr.deps
+            if DeclarationInfo in target
+        ],
     )
-    deps_runfiles = [d[DefaultInfo].default_runfiles for d in ctx.attr.deps]
+
+    npm_linked_deps_depsets = [
+        target[JsInfo].npm_linked_deps
+        for target in ctx.attr.deps
+        if JsInfo in target
+    ] + [
+        target[NpmLinkedPackageInfo].transitive_files
+        for target in ctx.attr.deps
+        if NpmLinkedPackageInfo in target
+    ]
+    for target in ctx.attr.deps:
+        if NpmLinkedPackagesInfo in target:
+            for package in target[NpmLinkedPackagesInfo].packages:
+                npm_linked_deps_depsets.append(package.transitive_files)
+    npm_linked_deps_depset = depset(transitive = npm_linked_deps_depsets)
 
     return [
+        JsInfo(
+            sources = sources_depset,
+            transitive_sources = transitive_sources_depset,
+            declarations = declarations_depset,
+            transitive_declarations = transitive_declarations_depset,
+            npm_linked_deps = npm_linked_deps_depset,
+            # npm_link_package_deps = ,
```

Comment:
?

Reply:
WIP, I'm working up to that one which will fix https://github.com/aspect-build/rules_ts/issues/96

Reply:
This is done now. See doctring in JsInfo for what it does.